### DISCUSSION TITLE
🙋Restore backup in a docker DSMR

### DISCUSSION THREAD

**Discussion by Rob99999** [dsmrreader/dsmr-reader#1593] [Questions & answers / How-to's] [answered]
### Description

Ik heb een docker container draaien met DSMR (in proxmox). Door een onzorgvuldige update (DSMR 4 naar 5 en changelog niet goed gelezen ) was de database niet meer juist. En vervolgens van de ene loophole in de andere. Uiteindelijk een geheel nieuwe install gerealiseerd die nu werkt. Nu ben ik mijn data kwijt. Gelukkig heb ik een back-up via Dropbox veiliggesteld. 

Na veel lezen ben ik nog onzeker hoe ik deze restore kan uitvoeren. Ik ben niet heel goed thuis in een linux omgeving. Heb wel  een bericht gelezen van [xirixiz](https://github.com/xirixiz) commented [on 1 Oct 2018](https://github.com/dsmrreader/dsmr-reader/issues/490#issuecomment-425857544) • waar volgens mij de juiste route is weergegeven.

Mijn SQL heeft echter een Gz. extensie. Moet ik deze eerst uitpakken. Moet ik de naam van de Sql db aanpassen naar het commando. Hoe krijg ik deze vervolgens in de juiste folder (in proxmox) in docker. en klopt het dat ik dan alleen nog de weergegeven commando's hoef uit te voeren?

### DSMR-reader version

DSMR-reader v5.1

### DSMR-reader platform

Docker (e.g. Xirixiz)

**Comment by dennissiemensma** [comment]
Bedankt voor je vraag. Je kunt dit het beste voorleggen bij of aan @xirixiz want ik draai zelf (nog) geen Docker en heb ook geen restore hoeven uit te voeren daarmee.

**Comment by xirixiz** [comment]
Hoi Rob, je kan de tgz idd uitpakken en dan de restore instructies volgen: 
```tar xvzf your_file.tar.gz```

**Comment by Rob99999** [comment]
Bedankt voor jullie antwoorden. Via WinSCP naar Proxmox en met PCT push sql bestand naar DSMR_backups folder geplaatst. @xirixiz ; het is geen tar.gz maar en sql.gz als ik het goed zie. Ik heb het bestand derhalve via 7zip in windows uitgepakt en vervolgens verplaatst. nu loop ik echter tegen het volgende issue aan:
Container gestopt en vervolgens 2de commando ingegeven
root@Docker ~/Compose# docker exec -t dsmrdb dropdb dsmrreader -U dsmrreader
Error: No such container: dsmrdb
root@Docker ~/Compose# dir
docker-compose.yml  dsmr_backups  dsmrdb  unifi
root@Docker ~/Compose# cd dsmrdb
root@Docker ~/Compose/dsmrdb# docker exec -t dsmrdb dropdb dsmrreader -U dsmrreader
Error: No such container: dsmrdb  

Er zou geen dsmrdb container zijn. ik begrijp de foutmelding niet.

**Comment by xirixiz** [comment]
Ah, ok, dan is het een gzip bestand, die kan je idd met gunzip of 7zip of een andere willekeurige tool uitpakken.
Welke Docker containers heb je draaien voor dsmr? In de voorbeelden die ik gebruik wordt er standaard een dsmrdb container aangemaakt, maar bij jou heeft die container een andere naam. Je moet de naam van de DSMR Docker database container achterhalen en dat gebruiken ipv dsmrdb.

Met ```docker ps -a``` kan je wel achterhalen welk id of naam de container heeft.

**Comment by Rob99999** [comment]
Ik heb wel de folder en kan het bestand erna uploaden. als ik dan DSMR stop en vervolgens het commando " docker exec -t dsmrdb createdb -O dsmrreader dsmrreader -U dsmrreader" invoer krijg ik een foutmelding dat de dsmrdb container aan het opstarten is. In de log krijg ik deze foutmelding:  

PostgreSQL Database directory appears to contain a database; Skipping initialization

2022-03-05 00:03:32.731 CET [1] FATAL:  database files are incompatible with server
2022-03-05 00:03:32.731 CET [1] DETAIL:  The data directory was initialized by PostgreSQL version 12, which is not compatible with this version 13.6.

**Comment by xirixiz** [comment, marked as answer]
In dat geval is de backup gemaakt in Postgres 12 en die probeer je nu te herstellen in Postgres 13. Het enige dat je kan doen is starten met Postgres 12, dan een restore uitvoeren en daarna upgraden naar Postgres 13.

**Comment by style2k** [comment]
ik heb nog een oude 4.x versie volgens mij op pi draaien maar wil ook op een docker gaan draaien zodat het updaten makkelijker te doen is enzo . waar moet ik op letten als ik naar een docker versie ga wat betreft de data ? is er een route waarmee ik makkelijk van een normale install naar een docker versie kan incl data ?? of is er mss in een nieuwere versie een optie om oude data in te laden ?

**Reply by xirixiz** [reply]
Docker kan je zien als een traditionele installatie, maar dan op een geisoleerde manier. 

Niet helemaal zo, maar om een idee te krijgen. Wanneer je postgres in Docker draait, dan draai je eigenlijk een mini os/vm met maar 1 functie/service, postgres in dit geval. Een Docker image is statisch, dus wanneer je een Docker container start met de betreffende image, dan kan je daar allerlei aanpassing op aanbrengen, maar wanneer je de Docker container dan weer stopt en vervolgens weer start, dan zijn alle wijzigingen verloren gegaan. Iets dat je uiteraard niet wilt in het geval van Postgres. In dat geval kan je dus bij het starten van een Docker container met het gebruik van een Docker image data/volumes bij het starten in de container mounten. In dat geval zijn dus het proces en de data van elkaar gescheiden.

Ik hoop dat het iets meer duidelijkheid geeft, maar in feite is er dus geen verschil wanneer je Postgres met of zonder Docker gebruikt, slechts de opzet wijkt onderling van elkaar af.

Dit gaat overigens specifiek om DSMR Reader en DSMR Reader icm Docker. Het is in jouw geval denk ik beter eerst wat meer kennis op te doen omtrent Docker zodat je beter begrijpt hoe het werkt en hoe je het kan gebruiken.

**Reply by style2k** [reply]
naja ik heb op de synology meer docker containers in gebruik en dat werkt opzich prima. het ding is ik ben niet zo van de linux command line dingen en het updaten van dsmr steeds is voor mij een heel gedoe steeds . daarom update ik niet zo vaak omdat het te veel gedoe is gewoon. en wat ik op me synology docker heb update ik automatisch via watch tower of evt in portainer, nieuwe image en klaar . nu wil ik dsmr dus ook in een docker draaien (wel op een pi natuurlijk) ook op de nas zijn docker containers idd aan een map gekoppeld waar alles in staat . 
maar de vraag is eigenlijk meer wat is de makkelijkste manier om mijn data van een gewone install over te zetten in een docker versie ? want volgens ij is het niet zo dat ik in dsmr kan zeggen load data oid tenminste niet in de versie die ik heb .
